Asynchronous Serial API fixes and refactoring

lrusinowicz opened the pull request:
### Description

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->

1. As RX and TX flows are separate on Serial device, read and write
   functionalities should be completely separate, including any deep
   sleep locking etc.
2. Aborting of asynchronous operation is necessarily hazardous, as
   operation can end in interrupt anywhere from the point of decision
   until it is actually aborted down the abort_...() method.
   Proper deep sleep unlocking requires then use of the critical
   section and should be contained completely within API implementation.
3. Because of point 2, API user should be able to call abort_..()
   operation independently of whether the operation is pending or not.
4. User may want to use asynchronous API without a callback (especially
   for write operations), for example in a command-response scheme
   end of write operation is usually meaningless. The intuitive
   method is to submit NULL pointer for a callback. For this reason
   depending on the _callback field in determining whether the
   operation is in progress seems to be uncertain.
5. Updated API documentation to cover above issues and to explicitly
   point out that any event ends the operation.


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [x] Fix
    [x] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers
 
<!-- 
    Optional
    Request additional reviewers with @username
-->

@0xc0170

ciarmcom commented:
@lrusinowicz, thank you for your changes.
@0xc0170 @ARMmbed/mbed-os-core @ARMmbed/mbed-os-maintainers please review.

cmonr commented:
@ARMmbed/mbed-os-hal @kjbracey-arm Please take a look at this since this PR modifies Serial behavior

kjbracey commented:
Afraid I don't feel terribly qualified to comment - I've never dared use the asynchronous serial API. Not sure who to suggest to cover it.

From what I can see though, this does look like a solid improvement.

cmonr commented:
@kjbracey-arm Thanks for the input.

Wondering if @ARMmbed/mbed-os-coresw could take a look at this as well?

0Grit commented:
> Afraid I don't feel terribly qualified to comment - I've never dared use the asynchronous serial API.

@kjbracey-arm Any particular reason why? I default to async api's in most cases, however I've all but confirmed that the async api's in Mbed OS are a 2nd priority.

0xc0170 commented:
It would be easier to review if the commit is split to few logical parts.

I can see from review:

- adding set flags for tx/rx
- proper locking addition
- updating docs

Changes are explained in the commit msg but it's hard to follow all in one.

lrusinowicz committed:
Asynchronous Serial API fixes and refactoring, part 1

1. As RX and TX flows are separate on Serial device, read and write
   functionalities should be completely separate, including any deep
   sleep locking etc.
2. User may want to use asynchronous API without a callback (especially
   for write operations), for example in a command-response scheme
   end of write operation is usually meaningless. The intuitive
   method is to submit NULL pointer for a callback. For this reason
   depending on the _callback field in determining whether the
   operation is in progress seems to be uncertain, so introduced
   additional flags for this purpose.

lrusinowicz committed:
Asynchronous Serial API fixes and refactoring, part 2

Aborting of asynchronous operation is necessarily hazardous, as
operation can end in interrupt anywhere from the point of decision
until it is actually aborted down the abort_...() method.
Proper deep sleep unlocking requires then use of the critical
section and should be contained completely within API implementation.

lrusinowicz committed:
Asynchronous Serial API fixes and refactoring, part 3

Updated API documentation to better explain API operation and to explicitly
point out that any event ends the operation.

cmonr reviewed (approved):
LGTM
